### DISCUSSION TITLE
Customers not always linked to order

### DISCUSSION THREAD

**Discussion by benoj** [saleor/saleor#8508] [Q&A] [answered]
Hi there,

I posted an [issue](https://github.com/saleor/saleor/issues/8464) on the repo but nothing back yet, so I'm hoping to get some ideas from folks here. I'm seeing some orders are not being linked to users when they purchase. Does anyone know why this may be?

### Screenshots 
here we see the customer has no orders in their customer profile (from admin UI)
<img width="571" alt="Screenshot 2021-10-26 at 17 35 19" src="https://user-images.githubusercontent.com/2669116/138922685-92230cb2-19b3-416f-987f-59ff6aa16d97.png">

Clicking the view orders button - I can see the orders connected to the email.
<img width="789" alt="Screenshot 2021-10-26 at 17 35 35" src="https://user-images.githubusercontent.com/2669116/138922678-d46b84c1-3626-44a1-af53-5d77951ee6bb.png">

**Saleor Version**
3.0.0-b.19

**Comment by karolkielecki** [comment]
@benoj Hi! Thanks for reporting. You mentioned that `some orders` / `not always` -- does it mean that some other orders are linked to customers? Could you please provide more information? Once I have more info, I'll pass it to the team.

**Comment by maarcingebala** [comment]
Sorry for not responding in the original issue, we had yet no time to try to reproduce the issue. <aybe the orders were created by an unauthenticated user, but with an email that is already registered as an account in Saleor, and the orders were not correctly matched with the account? When anonymous users create orders we look them up by email and connect to their account if I'm correct - maybe something is broken there.

**Comment by IKarbowiak** [comment]
@benoj, could you check what would you get for these orders on the field `user` from API? It seems like the `Customer` column is not fulfilled with the use of `user` value from `order` but based on order `userEmail` and with the use of customer search. I suppose that when `userEmail` fits the user from `customerSearch`, the customer name is shown.

**Comment by benoj** [comment]
@karolkielecki  - Yes SOME orders are connected some are not. In fact I'm looking at one user who has one order connected and one who has not connected to the same user.

**Comment by benoj** [comment]
@IKarbowiak 

I ran the following query:

```gql
query {
  orders(first: 10, ) {
    totalCount
    edges {
      node {
        number
        status
        userEmail
        user {
          email
        }
      }
    }
  }
}
```

here is a partially redacted response:

```json
{
  "data": {
    "orders": {
      "totalCount": 10,
      "edges": [
        {
          "node": {
            "number": "787902210",
            "status": "FULFILLED",
            "userEmail": "XXXXXXX@yahoo.com",
            "user": {
              "email": "XXXXXXX@yahoo.com"
            }
          }
        },
        {
          "node": {
            "number": "787902208",
            "status": "FULFILLED",
            "userEmail": "YYYYYYY@live.com",
            "user": null
          }
        },
        {
          "node": {
            "number": "6",
            "status": "FULFILLED",
            "userEmail": "XXXXXXX@yahoo.com",
            "user": null
          }
        }
      ]
    }
  }
}

```


So here we can see the user is null for one of user `XXXXXXX@yahoo.com`'s orders. Also user  `YYYYYYY@live.com` has an account but isn't linked?

**Comment by IKarbowiak** [comment]
Right now, when checkout is created, `user` is only assigned when the user is authenticated otherwise, only the `checkout.email` value is set. When `order` is created, the `user` and `userEmail` values are taken from `checkout`. So in your situation, I suppose that some orders were created by the logged user, and some of them only with the use of user `email` that exists in the database, but the user wasn't logged in during checkout.

I'm not sure if it's a good approach, maybe we should try to find the corresponding user when the order is created.

**Reply by benoj** [reply]
So I assumed based on this: https://docs.saleor.io/docs/2.11/developer/users#previous-orders

That orders are linked by email anyway?

**Comment by benoj** [comment]
So just looking through some of the checkout code.

We create our checkouts early when users add to basket (as we wanted them to be able to track them across devices) - this means when we initially create checkouts, we assign an anonymous unique email address. 

Looks like the CheckoutCreateInstance has the following:
```python
    @classmethod
    def get_instance(cls, info, **data):
        instance = super().get_instance(info, **data)
        user = info.context.user
        if user.is_authenticated:
            instance.user = user
        return instance
```

However, the checkout update email simply does this:

```python
    @classmethod
    def perform_mutation(cls, _root, info, email, checkout_id=None, token=None):
        # DEPRECATED
        validate_one_of_args_is_in_mutation(
            CheckoutErrorCode, "checkout_id", checkout_id, "token", token
        )

        if token:
            checkout = get_checkout_by_token(token)
        # DEPRECATED
        else:
            checkout = cls.get_node_or_error(
                info, checkout_id or token, only_type=Checkout, field="checkout_id"
            )

        checkout.email = email
        cls.clean_instance(info, checkout)
        checkout.save(update_fields=["email", "last_change"])
        info.context.plugins.checkout_updated(checkout)
        return CheckoutEmailUpdate(checkout=checkout)
```

So I'm thinking - for users who are NOT logged in and initiate a checkout in our system - the user attribute is never updated once they log-in. 

I imagine if I update the above to check the Users against the updated email (and they should have a token here?) then this should work?

**Reply by IKarbowiak** [reply]
Yes, I think it should work.
We should consider changing this behaviour. @maarcingebala

**Comment by benoj** [comment]
OK updating the above method to this resolved the issue - another option would be to perform this AFTER the checkout has been completed - will create a PR for this is this is something which will be valuable to other Saleor users?

```diff
   from ..utils import get_user_or_app_from_context
 from .types import Checkout, CheckoutLine
 from .utils import prepare_insufficient_stock_checkout_validation_error
+from ...account.models import Address, User
 
 ERROR_DOES_NOT_SHIP = "This checkout doesn't need shipping"
 
 
 if TYPE_CHECKING:
-    from ...account.models import Address
     from ...checkout.fetch import CheckoutInfo
 
 
@@ -973,6 +973,7 @@ class CheckoutEmailUpdate(BaseMutation):
 
     @classmethod
     def perform_mutation(cls, _root, info, email, checkout_id=None, token=None):
+        user = info.context.user
         # DEPRECATED
         validate_one_of_args_is_in_mutation(
             CheckoutErrorCode, "checkout_id", checkout_id, "token", token
@@ -986,9 +987,13 @@ class CheckoutEmailUpdate(BaseMutation):
                 info, checkout_id or token, only_type=Checkout, field="checkout_id"
             )
 
+        if user.is_anonymous or email != user.email:
+            user = User.objects.get(email=email)
+
+        checkout.user = user
         checkout.email = email
         cls.clean_instance(info, checkout)
-        checkout.save(update_fields=["email", "last_change"])
+        checkout.save(update_fields=["email", "user", "last_change"])
         info.context.plugins.checkout_updated(checkout)
         return CheckoutEmailUpdate(checkout=checkout)

```

**Reply by IKarbowiak** [reply]
This line: `user = User.objects.get(email=email)` will raise you an error when user with this `email` do not exist. Use `filter` and `first` instead.

**Comment by benoj** [comment]
@IKarbowiak I've created a PR with your feedback: https://github.com/saleor/saleor/pull/8518

**Comment by IKarbowiak** [comment]
@benoj After discussion, we decided to link created order to a user instance, if exists. This will require changes in `saleor/checkout/complete_checkout.py` in `_process_user_data_for_order` method. If you feel like it, you can try to make these changes; otherwise, we will change it as soon as possible. Thank you for reporting the issue.

**Comment by rolego** [comment]
Dear all

Even if I'm not sure if I was able to understand any detail of this discussion, I could not found an answer regarding the following scenario:

1. create an order anonymously with email address abc@xyz.com
2. register a customer with the same email address

Current behaviour in the Dashboard:
a) in the list of customers "No. of Orders" is 0 for this customer
b) in the detail view of this customer the list in "Recent Orders" is empty
c) pressing "View all orders" -> the (anonymously created) order is listed

Expected behaviour:
a) No of Orders > 0 
b) the order is shown in the list
c) ok

To my understanding this seems to be a bug and since this is an important user flow for us I would appreciate if this could get fixed.

**Reply by karolkielecki** [reply]
Hi @rolego, which version are you using?

**Reply by rolego** [reply]
@karolkielecki , I'm currently using a dev instance in the cloud. Assuming they all have the same version....

**Reply by patrys** [reply]
@rolego Follow-up question, did you confirm the email address after registration?

**Reply by rolego** [reply]
No. I just created the customer in the dashboard.

Background of all this: I will try to integrate auth0 with the OIDC plugin. 
With this test, I wanted to try out how saleor will behave in this scenario....

**Reply by rolego** [reply]
So, one of our scenarios for new customers looks like this:

1. customer buys anonymously 
2. onboarding in auth0
3. customer gets created in saleor and previously created order should be linked

**Reply by IKarbowiak** [reply]
Our current implementation requires email confirmation for linking the new user with existing orders. That's why orders wasn't linked.

**Reply by rolego** [reply]
OK, thanks for your answer.

Hopefully this somehow will work with OIDC....

**Comment by timuric** [comment, marked as answer]
Our current implementation requires email confirmation for linking the new user with existing orders. That's why orders weren't linked.
